python/tvm/relay/frontend/onnx.py

```diff
@@ -944,6 +946,36 @@ def _impl_v1(cls, inputs, attr, params):
         return Gelu._impl_v1([inp], attr, params)
 
 
+class LayerNormalization(OnnxOpConverter):
+    """Operator converter for LayerNormalization from Microsoft onnxruntime contrib opset."""
+
+    @classmethod
+    def _impl_v17(cls, inputs, attr, params):
+        x = inputs[0]
+        gamma = inputs[1]
+        beta = inputs[2]
+        axis = attr.get("axis", -1)
+        eps = attr.get("epsilon", 1e-5)
+        # according to the onnx doc, given the int axis (default -1)
+        # to compute the mean and inv_stdev which are of dim [d[0], ..., d[axis-1], 1, ..., 1]
+        # the actual computation is over (axis, ..., rank(x) - 1) axes
+        # see https://github.com/onnx/onnx/blob/main/docs/Changelog.md#layernormalization-17
+        rank = len(infer_shape(x))
+        axis = tuple(range(axis, rank)) if axis >= 0 else tuple(range(rank + axis, rank))
+        dtype = infer_type(x).checked_type.dtype
+        mean = _op.mean(x, axis, keepdims=True)
```

Comment:
I believe this is inefficient as mean will do a reduction along the axis and variance will do two reductions (one to find mean, one to find the mean of (X - E[X])^2. Giving three total reductions across the axis.

Instead just manually do E[(X - E[X])^2] to get two total reductions.

Reply:
Not sure if I understand completely! the test cases test for actual match between all outputs (with dims included) and the "wide" axes takes care of over what axes the mean and variance are computed which conform to what onnx spec is referring [here](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#layernormalization-17) where 

> the shape of Mean and InvStdDev is `[d[0], ..., d[axis-1], 1, ..., 1]`

Reply:
If we look at `python/tvm/relay/op/reduce.py` we see _op.variance calls `mean` which is superfluous. You can probably change it by giving it an option to pass in the mean.

Reply:
Isn't that (the point of rely) gets optimized over the simplification passes that rely does so we don't need to worry too much about it?

Reply:
Yes, common subexpression elimination pass should eliminate this. However, CSE can't do magic, and has limitations. You can probably check if it will be removed here by running this generated relay through the pass. Perhaps it will. 

In general, unless it adds a great amount of overhead in the reasoning of a piece of code, I would not duplicate ops however.

There are cases where it should but does not: https://github.com/apache/tvm/issues/12977

Reply:
Comment addressed. Please check again. Thanks!